[Requirement Machine] Diagnose redundant requirements.

hborla opened the pull request:
Plumb source location info through the rewrite system by storing the structural requirements used to derive explicit rules, and store the structural requirement ID in the rule itself. Structural requirement IDs are propagated from redundant rules to their replacements via `RewriteSystem::propagateRedundantRequirementIDs` after homotopy reduction. `RewriteSystem::getRedundantRequirements` computes the set of redundant requirement diagnostics to be emitted from `RequirementSignatureRequestRQM` and  `InferredGenericSignatureRequestRQM`.

This change does not (yet) emit notes to show where a requirement is made redundant.

Resolves: rdar://88135415

hborla committed:
[RequirementMachine] Preserve structural requirements during rule construction,
and diagnose trivially redundant rules in the rewrite system.

hborla committed:
[RequirementMachine] Store a requirement ID for explicit rules in the
rewrite system.

This ID can be used to index into the WrittenRequirements array in the
rewrite system to retrieve the structural requirement, e.g. for the
purpose of diagnostics.

hborla committed:
[RequirementMachine] Diagnose redundant requirements found during minimization
of the rewrite system.

hborla committed:
[RequirementMachine] Propagate explicit requirement IDs from redundant
rules in a separate pass after homotopy reduction.

RewriteSystem::propagateExplicitBits was too eager in propagating
IDs from explicit rules within rewrite loops, which resulted in bogus
redundancy warnings when the canonical form of an explicit rule was
given a different requirement ID. Instead, propagate requirement IDs
after homotopy reduction when redundant rules are computed and rewrite
loops are simplified.

hborla committed:
[RequirementMachine] Suppress redundancy warnings when an explicit, redundant
rule has a non-explicit, non-redundant rule in its rewrite path.

This fixes bogus redundancy diagnostics in cases where the canonical form
of a redundant rule is not explicit in source, e.g.

protocol Swappable2 {
  associatedtype A
  associatedtype B
  associatedtype Swapped : Swappable2
    where Swapped.B == A,
          Swapped.Swapped == Self
}

in the above case, the canonical rule for `Swapped.B == A` is the rule
[Swappable2:Swapped].[Swappable2:A] => [Swappable2:B], which is not
explicit.

hborla committed:
[RequirementMachine] Don't suppress redundancy diagnostics when the
non-explicit, non-redundant replacement rule is not in the rewrite
system's minimization domain.

hborla committed:
[RequirementMachine] Factor the 'rules once in empty context' computation out
into a method on RewritePath.

slavapestov commented on lib/AST/RequirementMachine/HomotopyReduction.cpp:
```diff
@@ -396,6 +410,51 @@ bool RewritePath::replaceRuleWithPath(unsigned ruleID,
+      if (!step.isInContext() && !evaluator.isInContext())
+        rulesInEmptyContext.insert(step.getRuleID());
+
+      ++ruleFrequency[step.getRuleID()];
```
Ok, I agree that 'frequency' is probably better than 'multiplicity' :)

slavapestov commented on lib/AST/RequirementMachine/HomotopyReduction.cpp:
```diff
@@ -656,6 +715,8 @@ void RewriteSystem::performHomotopyReduction(
     deleteRule(ruleID, replacementPath);
   }
+
+  propagateRedundantRequirementIDs();
```
Maybe this should be done once at the end? (performHomotopyReduction() is called three times)

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.h:
```diff
@@ -104,7 +104,12 @@ struct RuleBuilder {
   /// New rules derived from requirements written by the user, which can be
   /// eliminated by homotopy reduction.
-  std::vector<std::pair<MutableTerm, MutableTerm>> RequirementRules;
+  std::vector<std::tuple<MutableTerm, MutableTerm, Optional<unsigned>>>
```
I prefer to define a named type if I need a tuple of length > 2 because `std::get<N>()` is kind of ugly. Maybe RequirementRule or something?

slavapestov commented on lib/AST/RequirementMachine/RequirementLowering.h:
```diff
@@ -104,7 +104,12 @@ struct RuleBuilder {
+  std::vector<std::tuple<MutableTerm, MutableTerm, Optional<unsigned>>>
+      RequirementRules;
+
+  /// Requirements written in source code. The requirement ID in the above
```
These have been desugared at this point right? Might want to note that.

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.cpp:
```diff
@@ -705,10 +718,99 @@ void RewriteSystem::verifyRewriteRules(ValidityPolicy policy) const {
+  for (unsigned ruleID : indices(getRules())) {
+    auto &rule = getRules()[ruleID];
+
+    if (!rule.isExplicit() && !rule.isPermanent() && !rule.isRedundant() &&
```
isExplicit() doesn't imply getRequirementID() is not None though does it? Might be best to just use getRequirementID().hasValue() here

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.cpp:
```diff
@@ -705,10 +718,99 @@ void RewriteSystem::verifyRewriteRules(ValidityPolicy policy) const {
+
+    // If this rule is replaced using a non-explicit,
+    // non-redundant rule, it's not redundant.
+    auto rewritePath = redundantRules[ruleID];
```
Should RewriteSystem::RedundantRules be a MapVector then so that we can do lookups while preserving insertion order?

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.cpp:
```diff
@@ -705,10 +718,99 @@ void RewriteSystem::verifyRewriteRules(ValidityPolicy policy) const {
+      rulesPerRequirement;
+
+  // Collect non-explicit requirements that are not redundant.
+  llvm::SmallDenseSet<unsigned, 2> impliedRequirements;
```
Maybe nonExplicitNonRedundantRequirements? 'implied' is kind of vague (I know the comment explains it though)

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.cpp:
```diff
@@ -705,10 +718,99 @@ void RewriteSystem::verifyRewriteRules(ValidityPolicy policy) const {
+    auto ruleIDs = pair.second;
+
+    // If all rules derived from this structural requirement are redundant,
+    // then the requirement is unnecessary in the source code.
```
unless the replacement path for one of those redundant rules involves a non-redundant, non-explicit rule, right?

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.h:
```diff
@@ -44,6 +46,11 @@ class Rule final {
+  /// The written requirement ID, which can be used to index into the
+  /// \c WrittenRequirements array in the rewrite system to retrieve
+  /// the structural requirement.
+  Optional<unsigned> requirementID;
```
Can you squeeze this into the below bitfield? Use a biased representation where 0 means None and x+1 means Some(x). 16 bits ought to be enough

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.h:
```diff
@@ -165,9 +176,12 @@ class Rule final {
 
-  void markExplicit() {
-    assert(!Explicit && !Permanent &&
+  void markExplicit(Optional<unsigned> requirementID) {
```
Since propagateExplicitBits() is wholly separate from propagateRequirementIDs() can we split this up into markExplicit() and setRequirementID()?

The current isExplicit() bit is only used in the minimal conformances algorithm to break ties in some odd cases. I'd like to get rid of it one day.

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.h:
```diff
@@ -225,6 +239,12 @@ class RewriteSystem final {
+  /// The requirements written in source code.
+  std::vector<StructuralRequirement> WrittenRequirements;
+
+  /// The invalid requirements.
```
I'm worried we have too many vectors of RequirementErrors :)

This one is just redundancies detected from addExplicitRule() right? Perhaps addExplicitRule() should take a vector of RequirementErrors as an out parameter and push the errors there.

The request evaluator request would own the vector. It would first be used to collect errors from requirement desugaring. Then we can plumb it through the rule builder to collect the errors from addExplicitRule(), and then pass it down to getRedundantRequirements() to add the final set of redundancies.

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.h:
```diff
@@ -225,6 +239,12 @@ class RewriteSystem final {
   ArrayRef<const ProtocolDecl *> Protos;
 
+  /// The requirements written in source code.
+  std::vector<StructuralRequirement> WrittenRequirements;
```
In a similar vein to the Errors vector, I'm wondering if its better to avoid storing this in the rewrite system, and instead pass it down from the request evaluator request to getRedundantRequirements() since it's only used there to get the Requirement and SourceLoc for the diagnostic.

slavapestov commented on lib/AST/RequirementMachine/RequirementMachineRequests.cpp:
```diff
@@ -667,7 +681,20 @@ InferredGenericSignatureRequestRQM::evaluate(
       RequirementMachineMode::Enabled) {
-    hadError |= diagnoseRequirementErrors(ctx, errors, allowConcreteGenericParams);
+    // Diagnose invalid requirements dropped during desugaring.
+    hadError |= diagnoseRequirementErrors(ctx, errors,
```
This would be a bit cleaner if all errors were collected into a single vector owned by the request.

slavapestov commented on lib/AST/RequirementMachine/RewriteSystem.cpp:
```diff
@@ -705,10 +718,99 @@ void RewriteSystem::verifyRewriteRules(ValidityPolicy policy) const {
+
+  // Map redundant rule IDs to their rewrite path for easy access
+  // in the `isRedundantRule` lambda.
+  llvm::SmallDenseMap<unsigned, RewritePath> redundantRules;
```
I doubt it actually matters for efficiency but it's a bit unfortunate to copy the RewritePath here since it's backed by an std::vector. Since the property of the rewrite path that you use is whether it contains an 'implied rule', can you precompute that in the map value instead?

slavapestov reviewed (approved):
This looks great!

I have some suggestions for further cleanups, but only two I think are important:
- Untangling isExplicit() / setRequirementID() since they're now computed in separate places
- Replacing the std::tuple with a named type

For the rest it's up to you if you want to do them in this PR, another PR or not bother at all.

hborla committed:
[RequirementMachine] Instead of recording redundant requirements and other
diagnostics in the rewrite system, pass down the 'errors' vector from the
top-level requests.

hborla committed:
[RequirementMachine] Separate setting the explicit bit and the requirement
ID for a rewrite rule.

hborla committed:
[RequirementMachine] Avoid passing the requirement error vector through
initialization of the rewrite system.

Instead, the rewrite system can determine trivially redundant requirements
by finding structural requirements with no associated rewrite rules.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test Linux

hborla commented:
@swift-ci please smoke test Linux

hborla commented:
@swift-ci please smoke test Linux

hborla commented:
@swift-ci please smoke test Linux
